Formula/v8.rb

```diff
@@ -15,98 +17,79 @@ class V8 < Formula
-    # https://code.google.com/p/v8/issues/detail?id=3871
-    inreplace "tools/gyp/v8.gyp",
-              "'OTHER_LDFLAGS': ['-dynamiclib', '-all_load']",
-              "\\0, 'DYLIB_INSTALL_NAME_BASE': '#{opt_lib}'"
-
-    (buildpath/"build/gyp").install resource("gyp")
-    (buildpath/"third_party/icu").install resource("icu")
-    (buildpath/"buildtools").install resource("buildtools")
-    (buildpath/"base/trace_event/common").install resource("common")
-    (buildpath/"tools/swarming_client").install resource("swarming_client")
-    (buildpath/"testing/gtest").install resource("gtest")
-    (buildpath/"testing/gmock").install resource("gmock")
-    (buildpath/"tools/clang").install resource("clang")
-
-    system "make", "native", "library=shared", "snapshot=on",
-                   "console=readline", "i18nsupport=off",
-                   "strictaliasing=off"
-
-    include.install Dir["include/*"]
-
-    cd "out/native" do
-      rm ["libgmock.a", "libgtest.a"]
-      lib.install Dir["lib*"]
-      bin.install "d8", "mksnapshot", "process", "shell" => "v8"
+    ENV.prepend_path "PATH", buildpath
+
+    v8_version = head? ? "HEAD" : version
+    arch = MacOS.prefer_64_bit? ? "x64" : "x86"
+    output_name = "#{arch}.release"
+    output_path = "out.gn/#{output_name}"
+
+    # This env variable used by gclient to prevent depot_tools to update depot_tools on every call
+    # see https://www.chromium.org/developers/how-tos/depottools#TOC-Disabling-auto-update
+    ENV["DEPOT_TOOLS_UPDATE"] = "0"
+
+    # Initialize and sync gclient
+    system "gclient", "root"
+    system "gclient", "config", "--spec", <<~EOS
+      solutions = [
+        {
+          "url": "https://chromium.googlesource.com/v8/v8.git",
+          "managed": False,
+          "name": "v8",
+          "deps_file": "DEPS",
+          "custom_deps": {},
+        },
+      ]
+      target_os = [ "mac" ]
+      target_os_only = True
+    EOS
+    system "gclient", "sync", "-vvv", "-j", Hardware::CPU.cores, "-r", v8_version
+
+    # Enter the v8 checkout
+    cd "v8" do
+      gn_args = {
+        :target_os => "mac",
+        :target_cpu => arch,
+        :target_os_only => true,
+        :is_debug => false,
+        :is_component_build => true,
+        :v8_use_external_startup_data => false,
+        :v8_enable_i18n_support => true,
+      }.map do |k, v|
+        "#{k}=#{v.is_a?(String) ? "\\\"#{v}\\\"" : v}"
+      end.join(" ")
+
+      # Build with gn + ninja
+      system "gn gen \"#{output_path}\" --args=\"#{gn_args}\""
+      system "ninja", "-j", Hardware::CPU.cores, "-v", "-C", output_path, "d8"
+
+      # Install all the things
+      include.install Dir["include/*"]
+      cd output_path do
+        lib.install Dir["lib*.dylib"]
+
+        # Install d8 and icudtl.dat in libexec and symlink
+        # because they need to be in the same directory.
+        libexec.install Dir["d8", "icudt*.dat"]
+        bin.install_symlink libexec/"d8"
```

Comment:
I can confirm, that replacing this line with a wrapper script setting `--icu-data-file` will work:
```ruby
mkdir_p bin
(bin/"d8").atomic_write <<~EOS
  #!/bin/bash
  exec "#{libexec}/d8" --icu-data-file="#{libexec}/icudtl.dat" "$@"
EOS
```